### PR TITLE
SQL: (0.6) Pruning for VERSIONING partitions [closes #97]

### LINKED ISSUE #97: 0.6: partition pruning
SELECT should proactively restrict searched historical partition list. As each historical partition have min_time and max_time, SELECT should take into account these properties when choosing partitions to search. This is so-called *partition pruning*.

- [x] for system_time partitions must be strictly ordered: early VERSIONING partitions go first, AS OF NOW partition goes last;
- [x] historical partitions are filled from first to last, no need in DEFAULT marker;
- [x] when new VERSIONING partition is added it must be placed before AS OF NOW partition;
- [x] remove freelist;
- [x] assign correct values in `init_col_val()`, set `max_value` for AS OF NOW partition;
- [x] table-shared min/max stats;
- [x] update min/max on DML;
- [x] fix tests;
- [x] no need in item_expression for MAX_VALUE;
- [x] instead of partition max stat use next partition min stat for partition boundary;
- [x] udpate col_val before pruning.

### PR DESCRIPTION
**Pull request by midenok**
* based on RANGE pruning by COLUMNS (sys_trx_end) condition
* removed DEFAULT; AS OF NOW is always last; current VERSIONING as last non-empty (or first empty)
* Min/Max stats in TABLE_SHARE
* ALTER TABLE ADD PARTITION adds before AS OF NOW partition

### CODE REVIEW

File: sql/table.h [hunk 1]
```diff
@@ -731,26 +733,30 @@ struct TABLE_SHARE
   uint16 row_start_field;
   uint16 row_end_field;
   uint32 hist_part_id;
-  List<void> free_parts;
-  bool free_parts_init;
+  Vers_field_stats** stat_trx;
+  ulonglong stat_serial;
```
Comment (kevgs) [inline sql/table.h]: Why not making it an atomic? That will allow to avoid of using LOCK_stat_serial and result in less LOC.

Comment (midenok) [reply sql/table.h]: `LOCK_stat_serial` is **rwlock**, it protects not only `stat_serial`, but `stat_trx` data access -- see its usage.

Comment (kevgs) [reply sql/table.h]: In that case it would be great to have a comment with a list of variables this rwlock guards.

**Comment by kevgs** [comment]
Here is an unprotected access to TABLE_SHARE::stat_trx

```c++
#0  partition_info::vers_stat_trx (this=, fld=STAT_TRX_END, part_element_id=0) at /home/kevg/work/mariadb_10.2/sql/partition_info.h:494
#1  partition_info::vers_stat_trx (this=, fld=STAT_TRX_END, part=) at /home/kevg/work/mariadb_10.2/sql/partition_info.h:502
#2  partition_info::vers_interval_exceed (this=, max_time=1484318248, part=) at /home/kevg/work/mariadb_10.2/sql/partition_info.h:514
#3  vers_get_partition_id (part_info=, part_id=, func_value=) at /home/kevg/work/mariadb_10.2/sql/sql_partition.cc:3460
#4  get_parts_for_update (old_data="\361\001", new_data="\361\001", rec0="\361\001", part_info=, old_part_id=, new_part_id=, new_func_value=) at /home/kevg/work/mariadb_10.2/sql/sql_partition.cc:311
#5  ha_partition::update_row (this=, old_data="\361\001", new_data="\361\001") at /home/kevg/work/mariadb_10.2/sql/ha_partition.cc:4220
#6  handler::ha_update_row (this=, old_data="\361\001", new_data="\361\001") at /home/kevg/work/mariadb_10.2/sql/handler.cc:5949
#7  TABLE::delete_row (this=) at /home/kevg/work/mariadb_10.2/sql/sql_delete.cc:223
#8  mysql_delete (thd=, table_list=, conds=, order_list=, limit=18446744073709551615, options=0, result=) at /home/kevg/work/mariadb_10.2/sql/sql_delete.cc:635
#9  mysql_execute_command (thd=) at /home/kevg/work/mariadb_10.2/sql/sql_parse.cc:4523
#10 mysql_parse (thd=, rawbuf="delete from t1", length=14, parser_state=, is_com_multi=false, is_next_command=false) at /home/kevg/work/mariadb_10.2/sql/sql_parse.cc:7799
#11 dispatch_command (command=COM_QUERY, thd=, packet="delete from t1", packet_length=14, is_com_multi=false, is_next_command=false) at /home/kevg/work/mariadb_10.2/sql/sql_parse.cc:1804
#12 do_command (thd=) at /home/kevg/work/mariadb_10.2/sql/sql_parse.cc:1364
```

**Comment by midenok** [comment]
It is protected by `busy_rotation`. `LOCK_stat_serial` prevents statistics from being updated until `check_range_constants()` finishes (see `vers_update_range_constants()`).

File: sql/table.h (hunk 1 above)
Comment (midenok) [reply sql/table.h]: Fixed, but you should know that guard variables in most cases are meaningless alone. So you need to protect some critical sections of code.

**Review by kevgs — APPROVED** [review]

**Pull request merged**